makefiles/app_dirs.inc.mk

```
@@ -12,7 +12,18 @@ RIOTBASE ?= .
 APPLICATION_DIRS :=                   \
     fuzzing                           \
     bootloaders                       \
-    examples                          \
+    examples/advanced_examples        \
+    examples/essentials               \
+    examples/language_bindings/community_supported \
+    examples/language_bindings/officially_supported \
+    examples/networking/bluetooth_low_energy/misc_ble_examples \
+    examples/networking/bluetooth_low_energy/nimble \
+    examples/networking/coap          \
+    examples/networking/core_resource_directory \
+    examples/networking/dtls          \
+    examples/networking/gnrc_networking \
+    examples/networking/misc          \
+    examples/networking/mqtt          \
```

Does it make sense to use a command to scan for all subdirectories here? This way we would not need to touch this everytime there is a new category of example.

*Searching for the command...*

---

Sure, but unfortunately Makefile doesn't have support for `**` matching subdirectories :/

Also, there are Makefiles in some subfolders of examples which would probably also be catched by that. For example in `examples/essentials/subfolders/module`

---

(the most obvious is probably `$(shell find examples -type d)`, but I am trying to find a Make-based alternative)

---

The difficulty here is also that we want the _parent folder_ of all folders containing a Makefile (except special cases like the one mentioned above)

---

Yeah, the best Google came up with was also

```suggestion
  $(shell find examples -name Makefile -type f | xargs dirname | xargs dirname | sort -u) \
```

So let's use that (a bit safer than my original proposal as it only considers directories with `Makefiles` and then takes their parent)

---

Hum as expected:

```
$ find examples -name Makefile -type f | xargs dirname | xargs dirname | sort -u
examples/advanced_examples
examples/advanced_examples/bindist
examples/advanced_examples/pio_blink
examples/essentials
examples/essentials/subfolders
examples/language_bindings/community_supported
examples/language_bindings/community_supported/wasm
examples/language_bindings/officially_supported
examples/networking/bluetooth_low_energy/misc_ble_examples
examples/networking/bluetooth_low_energy/nimble
examples/networking/coap
examples/networking/core_resource_directory
examples/networking/dtls
examples/networking/gnrc_networking
examples/networking/misc
examples/networking/mqtt
```

We'd need to blacklist
```
examples/advanced_examples/bindist
examples/advanced_examples/pio_blink
examples/essentials/subfolders
examples/language_bindings/community_supported/wasm
```

---

Not sure if that's not even more brittle.

---

> We'd need to blacklist
> 
> ```
> examples/advanced_examples/bindist
> examples/advanced_examples/pio_blink
> examples/essentials/subfolders
> examples/language_bindings/community_supported/wasm
> ```

Huh? Why? If so, wouldn't they have been blacklisted before?

---

Ah got it... Mhhh

---

> Not sure if that's not even more brittle.

I'd argue not picking up applications that people then expect to be tested but actually aren't is more brittle than accidentally building modules (which will cause errors so easily detectable)... So I'd rather would go the blacklist route.

---

My proposal would be

```suggestion
    $(shell find examples -name Makefile -type f | \
      xargs dirname | xargs dirname | \
      grep -vFf $(RIOTBASE)/makefiles/app_dirs.blacklist | \
      sort -u) \
```

and then to create a file `makefiles/app_dirs.blacklist` with the content

```
examples/advanced_examples/bindist
examples/advanced_examples/pio_blink
examples/essentials/subfolders
examples/language_bindings/community_supported/wasm
```

---

If those shell commands are portable enough (I can't judge), I'd be fine with it. Should I do the same with the tests folder then?

---

This gives me more false positives within `bin` subfolders such as `tests/pkg/relic/bin/acd52832/pkg-build`. Probably something we should catch be excluding subdirectories of `bin`?

---

> If those shell commands are portable enough (I can't judge), I'd be fine with it. Should I do the same with the tests folder then?

They are all POSIX standard commands, so they _should_ work on any machine, but let me have a look, if grep supports `-F` (use plain strings for pattern instead of regex) and `-f` (take patterns from file) on restricted platforms.

---

> They are all POSIX standard commands, so they _should_ work on any machine, but let me have a look, if grep supports `-F` (use plain strings for pattern instead of regex) and `-f` (take patterns from file) on restricted platforms.

BusyBox `grep` supports both and to my information it is also supported on macOS, so I think both are safe to use. If you are not sure about `sort -u`, you could also use `sort | uniq`

---

I've used a single shell command now to list all application folders, please have a look :)